include/domain_bridge/service_bridge_impl.inc

```diff
@@ -0,0 +1,202 @@
+  auto serv_base_ptr = std::static_pointer_cast<rclcpp::ServiceBase>(serv);
+  node->get_node_services_interface()->add_service(serv_base_ptr, group);
+  return serv;
+}
+
+rclcpp::Node::SharedPtr
+get_node_for_domain(DomainBridgeImpl & impl, std::size_t domain_id);
+
+const std::string &
+get_node_name(const DomainBridgeImpl & impl);
+}  // namespace detail
+
+/// Bridge a service from one domain to another.
+/**
+ * \param service_bridge: Struct containing info about the service to bridge.
+ * \param options: Options for bridging the topic.
+ */
+template<typename ServiceT>
+void
+DomainBridge::bridge_service(
+  const TopicBridge & service_bridge,
+  const TopicBridgeOptions & options)
+{
+  const auto & node_name = detail::get_node_name(*impl_);
+  // Validate service name
+  const std::string & service_name = rclcpp::expand_topic_or_service_name(
+    service_bridge.topic_name, node_name, "/", true);
+
+  // If a remap name is provided, then validate it
+  // otherwise "remap" to the same name
+  std::string service_remapped = service_name;
+  if (!options.remap_name().empty()) {
+    service_remapped = rclcpp::expand_topic_or_service_name(
+      options.remap_name(), node_name, "/", true);
+  }
+
+  const std::size_t & from_domain_id = service_bridge.from_domain_id;
+  const std::size_t & to_domain_id = service_bridge.to_domain_id;
+
+  // Check if already bridged
+  if (service_bridge_impl_->bridged_services_.find(service_bridge) !=
+    service_bridge_impl_->bridged_services_.end())
+  {
+    std::cerr << "Service '" << service_name << "'" << " already bridged from domain " <<
+      std::to_string(from_domain_id) << " to domain " << std::to_string(to_domain_id) <<
+      ", ignoring" << std::endl;
+    return;
+  }
+
+  rclcpp::Node::SharedPtr from_domain_node = detail::get_node_for_domain(*impl_, from_domain_id);
+  rclcpp::Node::SharedPtr to_domain_node = detail::get_node_for_domain(*impl_, to_domain_id);
+
+  // Create a client for the 'from_domain'
+  auto client = from_domain_node->create_client<ServiceT>(
+    service_name,
+    rmw_qos_profile_services_default,
+    options.callback_group());
+
+  auto handle_request =
+    [client](
+      std::shared_ptr<detail::HackedService<ServiceT>> me,
+      const std::shared_ptr<rmw_request_id_t> & request_header,
+      const std::shared_ptr<typename ServiceT::Request> & request) -> void
+    {
+      // TODO(ivanpauno): What do we do if there's no service server available?
+      // This is a general problem in ROS 2, we don't have a clean way of communicating a service error ....
+      auto result = client->async_send_request(
+        request, [me = std::move(me), request_header](typename rclcpp::Client<ServiceT>::SharedFuture future_response)
+      {
+        // TODO(ivanpauno): Why do we get a future instead of the response directly here?
+        // That doesn't seem to make any sense.
+        // See if we can improve rclcpp service API.
```

Comment:
I agree. IIRC something similar was copied over to actions, but we later refactored it to remove the future.

Reply:
Yes, I will open a issues in rclcpp to discuss the problems found here.